pkg/forwarder/forwarder.go

```diff
@@ -40,85 +48,248 @@ func init() {
 	)
 }
 
-type Worker struct {
-	FromClient *metricsclient.Client
-	ToClient   *metricsclient.Client
-	Interval   time.Duration
-	Timeout    time.Duration
-	MaxBytes   int64
+// Config defines the parameters that can be used to configure a worker.
+// The only required fields are `From`, `ToAuthorize`, and `ToUpload`.
+type Config struct {
+	From          *url.URL
+	ToAuthorize   *url.URL
+	ToUpload      *url.URL
+	FromToken     string
+	ToToken       string
+	FromTokenFile string
+	ToTokenFile   string
+	FromCAFile    string
+
+	AnonymizeLabels   []string
+	AnonymizeSalt     string
+	AnonymizeSaltFile string
+	Debug             bool
+	Interval          time.Duration
+	LimitBytes        int64
+	Rules             []string
+	RulesFile         string
+	Transformer       metricfamily.Transformer
+}
 
-	from url.URL
-	to   *url.URL
+type Worker struct {
+	fromClient *metricsclient.Client
+	toClient   *metricsclient.Client
+	from       *url.URL
+	to         *url.URL
 
+	interval    time.Duration
 	transformer metricfamily.Transformer
 	rules       []string
 
-	lock        sync.Mutex
-	lastMetrics []*clientmodel.MetricFamily
+	lastMetricsLock sync.Mutex
+	lastMetrics     []*clientmodel.MetricFamily
+	rc              chan struct{}
+	reconfigureLock sync.Mutex
```

Comment:
suggestion: to keep things simpler, I suggest to reduce the amount of concurrency primitives for reloading to just one, namely the `rc` channel (which I guess is the "reconfigure" channel?).

Instead locking `Worker`'s internal state via `reconfigureLock` _and_ signalling the reload via `rc`, I suggest to remove the lock and communicate the need for reconfiguration via the `rc` channel only.

It could be of type `rc chan Config` and the desired new configuration would be received in `Run`'s select https://github.com/openshift/telemeter/pull/76/commits/41fd79ea65bc053054bb5cb1e660daed1117e653#diff-5dc0eebe45084d4d6def4a4bb1cbb0ddR285 block.

Then, instead of mutating `Worker`'s state in the calling go-routine, it would be mutated in the goroutine that executes the for loop in the `Run` method which would prevent the need for additional locking.

The `Reconfigure` method would be as simple as
```
func (w *Worker) Reconfigure(cfg Config) error {
	// Signal a restart to Run func.
	// This must be done in a goroutine to avoid potential deadlock.
	go func() { w.rc <- cfg }()
	return nil
}
```

Reply:
That is indeed a nice alternative. However, I see some issues with this solution is that make it not equivalent to the one in the commit as the reconfiguration would no longer be bounded by the Reconfigure func. Effectively, the reconfigure function would now be asynchronous but with the drawback of not notifying the result via a callback or promise. This means that calls to the /-/reconfigure endpoint will prematurely return a 200 before the reconfiguration has actually been applied. Also, it means that errors in a newly loaded configuration cannot be caught and exposed by the reconfigure function, and so they cannot be bubbled up via the HTTP response in a 4xx status code. Finally, it means that unit testing the reconfigure func is not really possible since there are no guarantees on the worker’s state after the function returns. This is why an additional synchronization point is required. We could achieve this in your suggestion by either: adding an additional channel to send back errors due to applying the new config or by changing the chan to accept a tuple of config + callback to notify a caller of the result of applying the new configuration. Neither of these options seems particularly preferable to the current solution. What do you think?

Reply:
It is also possible to collapse both locks into one single mutex to guard the entire state of the worker

Reply:
agreed, yes, a synchronous feedback of a (un)successful reconfiguration would be lost with my suggestion. But as this could also happen out of band we could still send a new instance of `Worker` via the channel.

I'll leave the final decision to you, but we should take care to lower the amount of concurrency primitives to a minimum.

I'm :+1: on collapsing at least `reconfigureLock` and `lastMetricsLock` into one.

Reply:
yes, sending the worker struct on the chan was my other thought, I thought it was a little bit hacky, but it would absolutely work :+1:

Reply:
I decided against using the channel to pass configuration to the worker. Instead, I chose to lock here, keep all the reconfiguration inside the Reconfigure func, and signal the Run func. Otherwise, if we instead passed a Config or Worker down the channel to the Run func, the Reconfigure func would not fully encapsulate a reconfiguration, meaning that any process that reconfigured a worker and then made a request against it, would be racing against the Run func. This keeps things deterministic and testable.